### PR TITLE
[MUL-289] Refresh payment method after update

### PR DESCRIPTION
**Pull request by DominikWolos**

### CODE REVIEW

File: features/shop/order.feature [hunk 1]
```diff
@@ -169,3 +169,17 @@ Feature: Spliting orders when cart was filled with products from different Vendo
     And I fill in "sylius_checkout_address[billingAddress][postcode]" with "Test code"
     And I submit form
     Then I should see "Envelope"
+
+  @ui
+  Scenario: Browsing orders history, can see selected shipping method
```
Comment (senghe) [inline features/shop/order.feature:174]:
```suggestion
  Scenario: Browsing orders history, can see selected payment method
```

File: features/shop/order.feature [hunk 2]
```diff
@@ -169,3 +169,17 @@ Feature: Spliting orders when cart was filled with products from different Vendo
...
+    And store has payment method "Cash on delivery" with code "cash_on_delivery"
+    And store has payment method "Bank transfer" with code "bank_transfer"
+    And I have 1 products in cart
+    And I finalize order with payment method "bank_transfer"
+    And I am on "/"
+    And I follow "My account"
+    And I follow "Order history"
+    Then I should see 1 orders
+    And I follow "Show"
+    Then I should see "Bank transfer"
```
Comment (senghe) [inline features/shop/order.feature:185]:
```suggestion
    Then I should see "Bank transfer" payment method
```

File: features/shop/order.feature [hunk 3]
```diff
@@ -169,3 +169,17 @@ Feature: Spliting orders when cart was filled with products from different Vendo
...
     And I submit form
     Then I should see "Envelope"
+
+  @ui
+  Scenario: Browsing orders history, can see selected shipping method
+    Given store has 2 products from same Vendor
+    And store has payment method "Cash on delivery" with code "cash_on_delivery"
+    And store has payment method "Bank transfer" with code "bank_transfer"
+    And I have 1 products in cart
+    And I finalize order with payment method "bank_transfer"
```
Comment (senghe) [inline features/shop/order.feature:179]: When from this place

File: features/shop/order.feature [hunk 4]
```diff
@@ -169,3 +169,17 @@ Feature: Spliting orders when cart was filled with products from different Vendo
...
+    And I am on "/"
+    And I follow "My account"
+    And I follow "Order history"
+    Then I should see 1 orders
+    And I follow "Show"
```
Comment (senghe) [inline features/shop/order.feature:184]:
```suggestion
    And I follow "Show" button
```

File: spec/Component/Order/Refresher/PaymentRefresherSpec.php [hunk 5]
```diff
@@ -31,18 +32,28 @@ public function it_is_initializable(): void
...
     ): void {
-        $order->recalculateItemsTotal()->shouldBeCalledOnce();
-        $order->recalculateAdjustmentsTotal()->shouldBeCalledOnce();
-        $order->getTotal()->willReturn(100);
+        $secondaryOrder->recalculateItemsTotal()->shouldBeCalledOnce();
+        $secondaryOrder->recalculateAdjustmentsTotal()->shouldBeCalledOnce();
 
-        $payments = new ArrayCollection([$payment->getWrappedObject()]);
-        $order->getPayments()->willReturn($payments);
+        $secondaryOrder->getTotal()->willReturn(100)->shouldBeCalledOnce();
```
Comment (senghe) [inline spec/Component/Order/Refresher/PaymentRefresherSpec.php:44]: Please don't mix stubs and mocks :pray:

File: src/Component/Order/Processor/SplitOrderByVendorProcessor.php [hunk 6]
```diff
@@ -34,8 +34,14 @@ public function __construct(
     public function process(OrderInterface $order): array
     {
         $isPrimaryOrder = $order->isPrimary() && 0 < $order->getSecondaryOrders()->count();
+
         if ($isPrimaryOrder) {
-            return [$order, ...$order->getSecondaryOrders()];
+            $orders = [$order, ...$order->getSecondaryOrders()];
+            foreach ($orders as $iOrder) {
```
Comment (senghe) [inline src/Component/Order/Processor/SplitOrderByVendorProcessor.php:40]: `$iOrder` us not a good name. I would recommend you:

```suggestion
            $secondaryOrders = [$order, ...$order->getSecondaryOrders()];
            foreach ($secondaryOrders as $secondaryOrder) {
```

File: src/Component/Order/Processor/SplitOrderByVendorProcessor.php [hunk 7]
```diff
@@ -58,8 +64,8 @@ public function process(OrderInterface $order): array
 
         $orders = [$order, ...$secondaryOrders];
 
-        foreach ($orders as $order) {
-            $this->paymentRefresher->refreshPayment($order);
+        foreach ($orders as $iOrder) {
```
Comment (senghe) [inline src/Component/Order/Processor/SplitOrderByVendorProcessor.php:67]: Same here. Please don't use $iOrder name

File: src/Component/Order/Refresher/PaymentRefresher.php [hunk 8]
```diff
@@ -27,10 +28,36 @@ public function refreshPayment(OrderInterface $secondaryOrder): void
...
-            $this->entityManager->persist($payment);
+
+        $this->refreshPaymentMethodAndAmount($secondaryOrder);
+
+        $this->entityManager->persist($secondaryOrder);
+    }
+
+    private function refreshPaymentMethodAndAmount(OrderInterface $secondaryOrder): void
+    {
+        $secondaryOrderPaymentCollection = $secondaryOrder->getPayments();
```
Comment (senghe) [inline src/Component/Order/Refresher/PaymentRefresher.php:39]: Just `$orderPayments` should be enough here. The context for this method should be independent from being primary/secondary order in my opinion

File: src/Component/Order/Refresher/PaymentRefresher.php [hunk 9]
```diff
@@ -27,10 +28,36 @@ public function refreshPayment(OrderInterface $secondaryOrder): void
...
+        $secondaryOrderPayment = $secondaryOrderPaymentCollection->last();
```
Comment (senghe) [inline src/Component/Order/Refresher/PaymentRefresher.php:40]: Why last? What if customer paid a part using giftcards, and a part using money?

File: src/Component/Order/Refresher/PaymentRefresher.php [hunk 10]
```diff
@@ -27,10 +28,36 @@ public function refreshPayment(OrderInterface $secondaryOrder): void
...
+
+        if (!$secondaryOrderPayment instanceof PaymentInterface) {
+            return;
+        }
+
+        $secondaryOrderPayment->setAmount($secondaryOrder->getTotal());
```
Comment (senghe) [inline src/Component/Order/Refresher/PaymentRefresher.php:46]: It's not exactly correct, as you can have multiple payments in single order.

**Review by senghe — CHANGES_REQUESTED** [review]

File: src/Component/Order/Refresher/PaymentRefresher.php (hunk 10 above)
Comment (DominikWolos) [reply src/Component/Order/Refresher/PaymentRefresher.php:46]: AFAIK multiple payments might occur after checkout process completes, but please correct me if I am wrong.

From my point of view - to split (primary) order between orders by vendor (secondary), we need a class that would take care of all changes at all points until checkout completes.

As you've pointed that payments might be multiple throughout Order lifecycle, it seems like we should secure this refresher from performing action on orders, for which we have completed the checkout.

Please let me know what you think.

File: spec/Component/Order/Refresher/PaymentRefresherSpec.php (hunk 5 above)
Comment (senghe) [reply spec/Component/Order/Refresher/PaymentRefresherSpec.php:43]:
```suggestion
        $secondaryOrder->getTotal()->willReturn(100);
```

File: src/Component/Order/Processor/SplitOrderByVendorProcessor.php [hunk 11]
```diff
@@ -77,4 +76,13 @@ private function vendorSecondaryOrderExits(array $secondaryOrders, ?VendorInterf
...
         return false;
     }
+
+    private function refreshPayments(array $orders): void
+    {
+        foreach ($orders as $order) {
+            $this->paymentRefresher->refreshPayment($order);
+        }
+
+        $this->entityManager->flush();
```
Comment (senghe) [inline src/Component/Order/Processor/SplitOrderByVendorProcessor.php:86]: I don't think the flush here is necessary

File: src/Component/Order/Refresher/PaymentRefresher.php (hunk 9 above)
Comment (senghe) [reply src/Component/Order/Refresher/PaymentRefresher.php:40]: Finally I think we can leave the topic of multiple payment methods per order.

File: src/Component/Order/Refresher/PaymentRefresher.php (hunk 8 above)
Comment (senghe) [reply src/Component/Order/Refresher/PaymentRefresher.php:39]: Okay, so please change this to `$secondaryOrderPayments` :pray:

File: src/Component/Order/Processor/SplitOrderByVendorProcessor.php (hunk 11 above)
Comment (DominikWolos) [reply src/Component/Order/Processor/SplitOrderByVendorProcessor.php:86]: After I dug up a bit, this processor is detached from OrderProcessor strategy, its only purpose and usage is in OrderController. Flush happens there via ResourceUpdateHandler. 
Removed flush from processor.

File: tests/Behat/Context/Setup/PaymentMethodContext.php [hunk 12]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+declare(strict_types=1);
```
Comment (senghe) [inline tests/Behat/Context/Setup/PaymentMethodContext.php:3]: Missing OS Header ;)

**Review by senghe — APPROVED** [review]
Approve in advance. Please add the OS header into the class, and then - merge :)

**Pull request merged**